### PR TITLE
Use simple array for AggregateException inner exceptions storage

### PR DESCRIPTION
**Pull request by marek-safar**
to drop unnecessary wrapping into ReadOnlyCollection. The change also simplified ctor chaining to avoid extra wrapping/casts where it's not needed.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/AggregateException.cs [hunk 1]
```diff
@@ -101,9 +101,7 @@ public AggregateException(string? message, Exception innerException)
         /// <exception cref="System.ArgumentException">An element of <paramref name="innerExceptions"/> is
         /// null.</exception>
         public AggregateException(string? message, IEnumerable<Exception> innerExceptions)
-            // If it's already an IList, pass that along (a defensive copy will be made in the delegated ctor).  If it's null, just pass along
-            // null typed correctly.  Otherwise, create an IList from the enumerable and pass that along.
-            : this(message, innerExceptions as IList<Exception> ?? (innerExceptions == null ? (List<Exception>)null! : new List<Exception>(innerExceptions)))
+            : this(message, innerExceptions == null ? Array.Empty<Exception>() : new List<Exception>(innerExceptions).ToArray(), cloneExceptions: false)
```
Comment (svick) [inline src/libraries/System.Private.CoreLib/src/System/AggregateException.cs:104]:
```suggestion
            : this(message, innerExceptions == null ? Array.Empty<Exception>() : innerExceptions.ToArray(), cloneExceptions: false)
```

Wouldn't this be simpler and more efficient?

Comment (marek-safar) [reply src/libraries/System.Private.CoreLib/src/System/AggregateException.cs]: That's linq extension which I think is not worth duplicating in SPC

File: src/libraries/System.Private.CoreLib/src/System/AggregateException.cs [hunk 2]
```diff
@@ -118,43 +116,29 @@ public AggregateException(string? message, IEnumerable<Exception> innerException
...
             if (innerExceptions == null)
             {
                 throw new ArgumentNullException(nameof(innerExceptions));
             }
 
-            // Copy exceptions to our internal array and validate them. We must copy them,
-            // because we're going to put them into a ReadOnlyCollection which simply reuses
-            // the list passed in to it. We don't want callers subsequently mutating.
-            Exception[] exceptionsCopy = new Exception[innerExceptions.Count];
+            _innerExceptions = cloneExceptions ? new Exception[innerExceptions.Length] : innerExceptions;
```
Comment (GrabYourPitchforks) [inline src/libraries/System.Private.CoreLib/src/System/AggregateException.cs:131]: `(Exception[])innerExceptions.Clone()`?

Comment (marek-safar) [reply src/libraries/System.Private.CoreLib/src/System/AggregateException.cs:131]: Is it better with the cast from object?

Comment (GrabYourPitchforks) [reply src/libraries/System.Private.CoreLib/src/System/AggregateException.cs:131]: Sorry, not sure what you mean by that question.

File: src/libraries/System.Private.CoreLib/src/System/AggregateException.cs (hunk 1 above)
Comment (svick) [reply src/libraries/System.Private.CoreLib/src/System/AggregateException.cs]: Sorry, I didn't realize that wasn't `Enumerable.ToArray()`.

**Comment by stephentoub** [comment]
What is the purpose of this change?

**Comment by marek-safar** [comment]
> What is the purpose of this change?

@stephentoub remove the unnecessary wrapping and ROS dependency.

**Comment by stephentoub** [comment]
But it still uses ROC:
```
private ReadOnlyCollection<Exception>? _rocView;
```
What am i missing?

**Comment by marek-safar** [comment]
It's used from a single property now which is rarely used (and could be avoided if needed). The linker will be able to remove the field by default whereas before it was a dependency from ctor which made it essentially non-trimmable.

**Comment by stephentoub** [comment]
> which is rarely used

Not really.  Most code paths that explicitly catch AggregateException then use InnerExceptions.  And even this PR is insufficient I'd expect to be able to ever trim it from corelib, since Task itself relies on InnerExceptions in a way that's unlikely to be trimmable.

If the goal is to enable ROC to be trimmable in most apps, let's make sure this PR actually does that.

**Comment by marek-safar** [comment]
> since Task itself relies on InnerExceptions in a way that's unlikely to be trimmable.

This didn't show up was wasm but I fixed the calls anyway to make this change more effective.

File: src/libraries/System.Private.CoreLib/src/System/AggregateException.cs (hunk 2 above)
Comment (marek-safar) [reply src/libraries/System.Private.CoreLib/src/System/AggregateException.cs:131]: I meant is Clone version better if it has to generate a cast check?

File: src/libraries/System.Private.CoreLib/src/System/AggregateException.cs [hunk 3]
```diff
@@ -21,15 +21,15 @@ namespace System
     [System.Runtime.CompilerServices.TypeForwardedFrom("mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089")]
     public class AggregateException : Exception
     {
-        private readonly ReadOnlyCollection<Exception> m_innerExceptions; // Complete set of exceptions. Do not rename (binary serialization)
+        private readonly Exception[] _innerExceptions; // Complete set of exceptions.
+        private ReadOnlyCollection<Exception>? _rocView;
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/AggregateException.cs:25]:
```suggestion
        private ReadOnlyCollection<Exception>? _rocView; // separate from _innerExceptions to enable trimming if InnerExceptions isn't used
```

File: src/libraries/System.Private.CoreLib/src/System/AggregateException.cs [hunk 4]
```diff
@@ -460,6 +399,8 @@ public override string ToString()
         ///
         /// See https://docs.microsoft.com/en-us/visualstudio/debugger/using-the-debuggerdisplay-attribute
         /// </summary>
-        private int InnerExceptionCount => InnerExceptions.Count;
+        internal int InnerExceptionCount => _innerExceptions.Length;
+
+        internal Exception[] IntenalInnerExceptions => _innerExceptions;
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/AggregateException.cs:404]:
```suggestion
        internal Exception[] InternalInnerExceptions => _innerExceptions;
```

**Review by stephentoub — APPROVED** [review]

**Pull request merged**